libp2p-swarm/src/connection_reuse.rs

```diff
@@ -0,0 +1,186 @@
+//!
+//! A `ConnectionReuse` can only be created from an `UpgradedNode` whose `ConnectionUpgrade`
+//! yields as `StreamMuxer`.
+//!
+//! # Behaviour
+//!
+//! The API exposed by the `ConnectionReuse` struct consists in the `Transport` trait
+//! implementation, with the `dial` and `listen_on` methods.
+//!
+//! When called on a `ConnectionReuse`, the `listen_on` method will listen on the given
+//! multiaddress (by using the underlying `Transport`), then will apply a `flat_map` on the
+//! incoming connections so that we actually listen to the incoming substreams of each connection.
+//! TODO: design issue ; we need a way to handle the substreams that are opened by remotes on
+//!       connections opened by us
+//!
+//! When called on a `ConnectionReuse`, the `dial` method will try to use a connection that has
+//! already been opened earlier, and open an outgoing substream on it. If none is available, it
+//! will dial the given multiaddress.
+//! TODO: this raises several questions ^
+//!
+//! TODO: this whole code is a dummy and should be rewritten after the design has been properly
+//!       figured out.
+
+use futures::{Future, Stream, Async, Poll};
+use futures::stream::Fuse as StreamFuse;
+use multiaddr::Multiaddr;
+use muxing::StreamMuxer;
+use smallvec::SmallVec;
+use std::io::Error as IoError;
+use transport::{Transport, ConnectionUpgrade, UpgradedNode};
+
+/// Allows reusing the same muxed connection multiple times.
+///
+/// Can be created from an `UpgradedNode` through the `From` trait.
+///
+/// Implements the `Transport` trait.
+#[derive(Clone)]
+pub struct ConnectionReuse<T, C>
+	where T: Transport,
+		  C: ConnectionUpgrade<T::RawConn>
+{
+	// Underlying transport and connection upgrade for when we need to dial or listen.
+	inner: UpgradedNode<T, C>,
+}
+
+impl<T, C> From<UpgradedNode<T, C>> for ConnectionReuse<T, C>
+	where T: Transport,
+		  C: ConnectionUpgrade<T::RawConn>
+{
+	#[inline]
+	fn from(node: UpgradedNode<T, C>) -> ConnectionReuse<T, C> {
+		ConnectionReuse { inner: node }
+	}
+}
+
+impl<T, C> Transport for ConnectionReuse<T, C>
+	where T: Transport + 'static, // TODO: 'static :(
+		  C: ConnectionUpgrade<T::RawConn> + 'static, // TODO: 'static :(
+		  C: Clone,
+		  C::Output: StreamMuxer + Clone,
+		  C::NamesIter: Clone // TODO: not elegant
+{
+	type RawConn = <C::Output as StreamMuxer>::Substream;
+	type Listener = ConnectionReuseListener<
+		Box<
+			Stream<
+				Item = (C::Output, Multiaddr),
+				Error = IoError,
+			>,
+		>,
+		C::Output,
+	>;
+	type Dial = Box<Future<Item = Self::RawConn, Error = IoError>>;
+
+	fn listen_on(self, addr: Multiaddr) -> Result<(Self::Listener, Multiaddr), (Self, Multiaddr)> {
+		let (listener, new_addr) = match self.inner.listen_on(addr.clone()) {
+			Ok((l, a)) => (l, a),
+			Err((inner, addr)) => {
+				return Err((ConnectionReuse { inner: inner }, addr));
+			}
+		};
+
+		let listener = ConnectionReuseListener {
+			listener: listener.fuse(),
```

Comment:
Why do you `fuse` here? It's not a safety issue and it's expected that misbehaving streams will cause logic bugs.

Reply:
The `ConnectionReuseListener` wraps around the listener and all the active connections. Every time either the listener or one the connections "triggers", we re-poll everything. Because of that, even if the `listener` finishes, we may continue to poll it. This can be problematic and [the documentation explicitly warns about this](https://docs.rs/futures/0.1.17/futures/stream/trait.Stream.html#panics).